types/priv_validator/socket.go

```diff
@@ -0,0 +1,273 @@
+		pvsc.conn.Close()
+	}
+}
+
+func (pvsc *PrivValidatorSocketClient) Address() data.Bytes {
+	pubKey := pvsc.PubKey()
+	return pubKey.Address()
+}
+
+func (pvsc *PrivValidatorSocketClient) PubKey() crypto.PubKey {
+	res, err := readWrite(pvsc.conn, PubKeyMsg{})
+	if err != nil {
+		panic(err)
+	}
+	return res.(PubKeyMsg).PubKey
+}
+
+func (pvsc *PrivValidatorSocketClient) SignVote(chainID string, vote *types.Vote) error {
+	res, err := readWrite(pvsc.conn, SignVoteMsg{Vote: vote})
+	if err != nil {
+		return err
+	}
+	*vote = *res.(SignVoteMsg).Vote
+	return nil
+}
+
+func (pvsc *PrivValidatorSocketClient) SignProposal(chainID string, proposal *types.Proposal) error {
+	res, err := readWrite(pvsc.conn, SignProposalMsg{Proposal: proposal})
+	if err != nil {
+		return err
+	}
+	*proposal = *res.(SignProposalMsg).Proposal
+	return nil
+}
+
+func (pvsc *PrivValidatorSocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat) error {
+	res, err := readWrite(pvsc.conn, SignHeartbeatMsg{Heartbeat: heartbeat})
+	if err != nil {
+		return err
+	}
+	*heartbeat = *res.(SignHeartbeatMsg).Heartbeat
+	return nil
+}
+
+//---------------------------------------------------------
+
+// PrivValidatorSocketServer implements PrivValidator.
+// It responds to requests over a socket
+type PrivValidatorSocketServer struct {
+	cmn.BaseService
+
+	conn        net.Conn
+	proto, addr string
+	listener    net.Listener
+
+	privVal PrivValidator
+	chainID string
+}
+
+func NewPrivValidatorSocketServer(logger log.Logger, socketAddr, chainID string, privVal PrivValidator) *PrivValidatorSocketServer {
+	proto, addr := cmn.ProtocolAndAddress(socketAddr)
+	pvss := &PrivValidatorSocketServer{
+		proto:   proto,
+		addr:    addr,
+		privVal: privVal,
+		chainID: chainID,
+	}
+	pvss.BaseService = *cmn.NewBaseService(logger, "privValidatorSocketServer", pvss)
+	return pvss
+}
+
+func (pvss *PrivValidatorSocketServer) OnStart() error {
+	if err := pvss.BaseService.OnStart(); err != nil {
+		return err
+	}
+	ln, err := net.Listen(pvss.proto, pvss.addr)
+	if err != nil {
+		return err
+	}
+	pvss.listener = ln
+	go pvss.acceptConnectionsRoutine()
+	return nil
+}
+
+func (pvss *PrivValidatorSocketServer) OnStop() {
+	pvss.BaseService.OnStop()
+	if err := pvss.listener.Close(); err != nil {
+		pvss.Logger.Error("Error closing listener", "err", err)
+	}
+
+	if err := pvss.conn.Close(); err != nil {
+		pvss.Logger.Error("Error closing connection", "conn", pvss.conn, "err", err)
+	}
+}
+
+func (pvss *PrivValidatorSocketServer) acceptConnectionsRoutine() {
+	for {
+		// Accept a connection
+		pvss.Logger.Info("Waiting for new connection...")
+		var err error
+		pvss.conn, err = pvss.listener.Accept()
+		if err != nil {
+			if !pvss.IsRunning() {
+				return // Ignore error from listener closing.
+			}
+			pvss.Logger.Error("Failed to accept connection: " + err.Error())
+			continue
+		}
+
+		pvss.Logger.Info("Accepted a new connection")
+
+		// read/write
+		for {
```

Comment:
only one connection at a time?

Reply:
@ebuchman Is one connection at a time by design, or should the server be able to hanlde multiple, if so are there any upper bounds we want to enforce, or any other limiting requirements?